pkg/man/eliomc.1
.\"                                      Hey, EMACS: -*- nroff -*-
.TH ELIOMC 1 2012-02-15
.SH NAME
eliomc, eliomcp, eliomopt, js_of_eliom, eliomdep \- the eliom build tools
.SH SYNOPSIS
.B eliomc
.RI [ options ]
.RI files

.B eliomcp
.RI [ options ]
.RI files

.B eliomopt
.RI [ options ]
.RI files

.B eliomdep
.RI [ options ]
.RI files

.B eliomdoc
.RI [ options ]
.RI files

.B eliompp
.RI [ -client | -server ]
.RI files

.B js_of_eliom
.RI [ options ]
.RI files
\-o
.RI filename.js
.SH DESCRIPTION
.B eliomc,
.B eliomcp
and
.B eliomopt
are wrapper around the OCaml compiler that ease the compilation of the
server part of projects based on the Eliom framework. They
respectively accept the same set of option than the
.BR ocamlc (1), ocamlcp (1)
and
.BR ocamlopt (1)
compilers plus the specific ones described in the
.I OPTIONS
section.

.B js_of_eliom
is a wrapper around the OCaml and Js_of_ocaml compilers that ease the
compilation of the client part of projects based on the Eliom
framework. It accepts the same set of option than the
.BR ocamlc (1)
compiler plus the specific ones described in the
.I OPTIONS
section.

.B eliomdep
is a wrapper around
.BR ocamldep (1)
that handles dependencies of an .eliom source files. It accepts the
same set of options plus the specific ones described in the
.I OPTIONS
section. It makes the assumption that server-side compiled modules
of OCaml files located in this directory subtree are in directory
.I _server
(with same subdirectories structure) (resp. in directory
.I _client
for client-side compiled modules).

.B eliompp (EXPERIMENTAL)
is a preprocessor which remove specific sections, depending on the given option.
The option
.BR -client
will remove all the top-level parts and server sections
and the option
.BR -server
will remove only the client sections.

.B eliomdoc (EXPERIMENTAL)
is a wrapper around the OCaml documentation generator
.B ocamldoc.
It allows you to generate documentation from
.B client
or
.B server
side. It accepts the same set of option than
.BR ocamldoc (1).
It automatically extracts the commentaries from the desired sections.
.B eliomdoc
is currently experimental, and you could find some undefined behaviours.
(see http://ocsigen.org/eliom/manual/workflow-compilation). Some well
known bugs are:
.br
.IP \[bu]
your files should always begin with a value and not with a comment. Otherwise,
camlp4 won't output the commentaries.
.br
.IP \[bu]
sometimes, comment node are not attached as expected. That's because
camlp4 (sometimes) remove extra new line between value elements.

.SH COMPILING ELIOM SOURCE FILES

The compilation of files with a .eliom extension is achieved in three
steps: infer the type of value sent by the server to the client;
compile the server part of the code and compile the client part. The
first two steps can be realised with
.B eliomc
and the last one with
.B js_of_eliom.

Both tools produce a .cmo file named as the original .eliom file. To
avoid overwriting the .cmo representing the server part with .cmo of
the client part, files generated by
.BR eliomc \ or \ eliomopt
are stored by default in a subdirectory named _server and files generated by
.B js_of_eliom
are stored in a subdirectory named _client. Those default directories
could be respectively overridden by the environment variables
ELIOM_SERVER_DIR and ELIOM_CLIENT_DIR.

The types inferred by
.B eliomc
for values sent by the server to the client are stored in an
intermediate files named as the original .eliom file and whose
extension is .type_mli. That file is required by
.B js_of_eliom
for compiling the client part of the .eliom file. The
.B eliomdep
tool correctly generate dependencies that intermediate file for the
server .cmo and the client .cmo.

.SH OPTIONS
.BR \-eliom-inc\ <dir>
Add <dir> to the list of eliom include directories (eliomdep only).
.TP
.TP
.BR \-dir\ <dir>
Specify the target directory for generated files
.TP
.BR \-package\ <name>
This is the same option as the ocamlfind one.
.TP
.BR \-predicates\ <p>
This is the same option as the ocamlfind one.
.TP
.BR \-no-autoload
Do not load commonly used syntax extensions
(ppx_deriving, lwt, js_of_ocaml, tyxml).
.TP
.BR \-type-conv
Use type_conv syntax extensions instead of deriving one. It has no effect
if used in conjunction with \-no-autoload.
.TP
.BR \-ppopt\ <opt>
Append <opt> to preprocessor invocation.
.TP
.BR \-jsopt\ <opt>
Append <opt> to js_of_ocaml invocation (js_of_eliom only).
.TP
.BR \-infer\ <opt>
For .eliom file, only generate the intermediate .type_mli file (eliomc and eliomopt only).
.TP
.BR \-noinfer\ <opt>
For .eliom file, do not generate the intermediate .type_mli file (eliomc and eliomopt only).
.TP
.BR \-help \ or \ \-\-help
Display a short usage summary and exit.
.SH EXAMPLES
The compilation of an Eliom projects composed of a server specific file named server.ml, a client specific file named client.ml and two common files name base.eliom and main.eliom, could be achieved with the following commands:

\ \ \ \ eliomc \-a \-o appl.cma server.ml base.eliom main.eliom
.br
\ \ \ \ js_of_eliom \-o appl.js client.ml base.eliom main.eliom

To avoid recompiling the whole project each times, this could be split in multiple steps:

\ \ \ \ eliomc \-c server.ml
.br
\ \ \ \ eliomc \-c base.eliom
.br
\ \ \ \ eliomc \-c main.eliom
.br
\ \ \ \ eliomc \-a \-o appl.cma _server/server.cmo _server/base.cmo _server/main.cmo
.br
\ \ \ \ js_of_eliom \-c client.ml
.br
\ \ \ \ js_of_eliom \-c base.eliom
.br
\ \ \ \ js_of_eliom \-c main.eliom
.br
\ \ \ \ js_of_eliom \-o appl.js _client/client.cmo _client/base.cmo _client/main.cmo



.SH SEE ALSO
.BR ocamlc (1),
.BR ocamlopt (1),
.BR js_of_ocaml (1),
.BR ocamlfind (1),
.BR ocamlcp (1),
.BR ocamldep (1).
.SH AUTHOR
eliomc, js_of_eliom, eliomdep and eliomopt were written by
Gregoire Henry <gregoire.henry@pps.jussieu.fr>.

eliomdoc was written by
Charly Chevalier <charly.n.chevalier@gmail.com>.
.PP
This manual page was written by Pierre Chambart <chambart@crans.org>.


CHANGES
===== 6.7 (2018-02-08) =====

* Eliom_client.Page_status: function while_active; ?stop parameter
* DOM caching: treat history as linear and keep only reachable pages in cache

===== 6.6 (2018-01-17) =====

* Add missing function Eliom_request_info.get_remote_port
* Eliom_client.Page_status.oninactive
* Some documentation on DOM caching
* Eliom_wrap: fix compilation on 32-bit architectures

===== 6.5 (2018-12-18) =====

* Compatibility with tyxml.4.3.x and js_of_caml.3.3.x

===== 6.4 (2018-12-17) =====

* Compatibility with Lwt 4.x
* New configuration option to set html content type
* Comet: deactivate when window does not have the focus; deactivate polling when idle
* eliom_request_info.in_request_handler
* Client-side non-localized parameters
* Provide a way to serve global data separately from the initial page

===== 6.3 (2018-02-08) =====

* Compatibility with OCaml 4.06, js_of_ocaml 3.0 and Lwt 3.0
* DOM caching (preview)
* Various small fixes and improvements

===== 6.2 (2017-02-07) =====

* Implement Eliom_client.onchangepage
* Fix URL-related bugs arising in mobile apps

===== 6.1 (2016-12-20) =====

* Remove redundant dependency on calendar
* Permit client routing to "./"
* -y parameter for eliom-distillery (do not ask)
* Eliom_client: do not execute onload after OCaml services
* Permit suffix params in Eliom_service.create_attached_post

===== 6.0 (2016-12-08) =====

* Improve Eliom_service and Eliom_registration APIs using GADTs
* Implement client-side services, useful for mobile apps. This includes
** client-side service registration (Eliom_registration)
** client-side service routing (Eliom_client.change_page_uri)
* Transition to PPX internally
* Compatibility with
** OCaml 4.03.0 and 4.04.0
** Js_of_ocaml 2.8.2 and newer
** TyXML 4.0 and newer
* Various bugfixes and improvements

===== 5.0 (2016-01-28) =====

* Add Eliom_shared and server-side Eliom_content.{Html5,Svg}.R modules
* Add PPX syntax extension
* Clean-up form APIs, now available under Eliom_content.Html5.{D,F}.Form
* Patches to significantly reduce the size of request data
* Compatibility with TyXML 3.6, Js_of_ocaml 2.7, and reactiveData 0.2
* Various bugfixes and enhancements

===== 4.2 (2015-07-21) =====

* Add Manip.children, to get the html children of an element.
* Simplify Html5 and Svg signatures using tyxml's signature functors.
* Various logging improvements, in particular in Eliom_client.
* Fix eliomdep's exit code when calling option "-sort".
* Fix #168: call_ocaml_service always sends boolean true.
* Makes server function return types covariant.
* Restore compatibility with ocsigenserver 2.6 and lwt 2.4.7.
* Various bugfixes and wiki updates.

===== 4.1 (2014-10-01) =====

 * Compatibility with js_of_ocaml-2.5 and tyxml-3.3.0
 * Improvements:
 ** Simplify typing of services
 ** Use GADT to encode service parameters (Eliom_parameter)
    Eliom need OCaml 4.00 at least.
 ** bugfixes for service parameters
 ** Various bugfixes and enhancement

===== 4.0 (2014-05) =====


 * Main new features:
 ** Reactive DOM elements (thanks to a contribution by besport.com)
 ** PUT and DELETE services for RESTful Eliom Applications
 (Contribution by Domoco)
 ** EXPERIMENTAL: Mobile apps with Eliom. By making it possible to start
 the application from client-side.
 ** Documentation improvements

 * Improvements:
 ** Improvements in typing of services (Contribution by Jacques-Pascal Deplaix)
 ** Filtering data from Eliom buses
 ** "caml" in value or type names now replaced by "OCaml" everywhere
 ** New module {{{Eliom_registration.File_ct}}} to make easier to send files with
 their content type
 ** Now possible to define services in shared sections
 ** Adding function {{{Eliom_client.change_url}}} to change the URL without doing a request
 ** Adding function {{{Eliom_service.attach_coservice'}}} to attach a non-attached coservice to an URL
 ** Improving comet configurations: now possible to stay idle instead of being completely inactive
 ** Now using defaultprotocol option (from ocsigenserver) to compute URLs
 ** The default expiration date for cookies on client side is now 10 years
 ** Now possible to send files in {{{Eliom_client.call_ocaml_service}}}
 ** Various additions and bugfixes in {{{Manip}}}, {{{Eliom_request_info}}},
 {{{Eliom_content.html5}}} (events) (contributions by besport.com)
 ** eliom-destillery renamed into eliom-distillery
 ** Templates for eliom-distillery
 ** New programs: eliomdoc and eliompp
 ** Ocamlbuild package
 ** Various bugfixes

And many other things (see git commits)

===== 3.0.0 (2012-12-10) =====

 * Language
 ** Generalized client values in server code
 ** Injections into client sections

 * Tools
 ** Added eliom-destillery for generating project scaffolds
 ** Support Eliom interface files (.eliomi) in eliomc, eliomdep
 ** eliomdep: Generate dependencies between eliom-modules
 ** eliomc: infer only with option -infer, drop option -noinfer
 ** eliomc: Basic support for -i on eliom-files
 ** eliom{c,dep,opt},js_of_eliom: -dump to output the intermediate code
 ** eliomc,js_of_eliom: always open Eliom_pervasives in eliom files

 * API
 ** Eliom_pervasives.server_function to easily access the from the client
 ** Get current state of a scope
 ** Module to access values of Eliom references in external states (Eliom_state.Ext)
 ** Scope names are now named scope hierarchies
 ** Iterate the scope hierarchy (group > session > client > request)
 ** Adding Eliom_parameter.(type_checker, neopt)
 ** Add functions to insert html5 in atom feeds
 ** Eliom_tools.{F,D}.html to ease creation of head-tag
 ** Eliom_tools.wrap_handler as an easy alernative to Eliom_registration.Customize
 ** Test for initial request of a client on the server

 * Changed server behaviour
 ** Eliom_state.discard_everything now also discards request state
 ** Don't send nodes as data when they are part of the document

 * Changed client behaviour
 ** Show progress cursor during navigation with change_page

src/tools/eliomc.ml

open Utils

let force_link_all = ref true

let link_all () =
  if !force_link_all
  then ["-linkall"]
  else []

let usage () =
  Printf.eprintf "Usage: %s <options> <files>\n" (Filename.basename Sys.argv.(0));
  Printf.eprintf "SPECIFIC OPTIONS:\n";
  Printf.eprintf "  -package <name>\tRefer to package when compiling\n";
  Printf.eprintf "  -no-autoload\t\tDo not load commonly used syntax extensions (deriving, lwt, js_of_ocaml, tyxml)\n";
  Printf.eprintf "  -type-conv\t\tUse type_conv syntax extensions instead of deriving one\n";
  Printf.eprintf "\t\t\tIt has no effect if used in conjunction with -no-autoload\n";
  if !kind <> `Client then
    Printf.eprintf "  -infer\t\tOnly infer the type of values sent by the server\n";
  Printf.eprintf "  -dir <dir>\t\tThe directory for generated files (default %S)\n"
    (if !kind = `Client then default_client_dir else default_server_dir);
  Printf.eprintf "  -type-dir <dir>\tThe directory to read server type files from (default %S)\n"
    default_type_dir;
  Printf.eprintf "  -server-types-ext <ext> The extension for server type files (default %S)\n"
    default_server_types_ext;
  if !kind = `Client then
    Printf.eprintf "  -jsopt <opt>\t\tAppend option <opt> to js_of_ocaml invocation\n";
  Printf.eprintf "  -ppopt <p>\t\tAppend option <opt> to preprocessor invocation\n";
  Printf.eprintf "  -predicates <p>\tAdd predicate <p> when resolving package properties\n";
  Printf.eprintf "  -ppx";
  Printf.eprintf "\t\t\tUse the PPX Eliom syntax extension (default: Camlp4)\n";
  Printf.eprintf "\t\t\tThe above description only applies to the first\n";
  Printf.eprintf "\t\t\tappearance of -ppx in the list of flags. Subsequent\n";
  Printf.eprintf "\t\t\tappearances require an argument and specify a\n";
  Printf.eprintf "\t\t\tPPX preprocessor to use (see STANDARD OPTIONS).\n";
  if !kind = `Client then
    Printf.eprintf "  -dont-force-linkall\t\tDo not add linkall option by default\n";
  create_filter !compiler ["-help"] (help_filter 2 "STANDARD OPTIONS:");
  if !kind = `Client then
    create_filter !js_of_ocaml ["-help"] (help_filter 1 "JS_OF_OCAML OPTIONS:");
  exit 1

(** Context *)

let jsopt : string list ref = ref []
let output_name : string option ref = ref None

type mode = [ `Link | `Compile | `Infer | `Library  | `Pack | `Obj | `Shared | `Interface ]

let mode : mode ref = ref `Link

let do_compile () = !mode <> `Infer
let do_interface () = !mode = `Interface
let do_dump = ref false

let create_process ?in_ ?out ?err ?on_error name args =
  wait ?on_error (create_process ?in_ ?out ?err name args)

let rec check_or_create_dir name =
  if name <> "/" then
    try ignore(Unix.stat name) with Unix.Unix_error _ ->
      check_or_create_dir (Filename.dirname name);
      try Unix.mkdir name 0o777 with
      (* this append sometime with // compilation *)
      | Unix.Unix_error (Unix.EEXIST,_,_) -> ()
      | Unix.Unix_error (_,_,_) ->
        Printf.eprintf "Unexpected error while creating directory %S" name

let prefix_output_dir name =
  match !build_dir with
  | "" -> name
  | d -> d ^ "/" ^ name

let chop_extension_if_any name =
  try Filename.chop_extension name with Invalid_argument _ -> name

let output_prefix ?(ty = false) name =
  let name =
    match !output_name with
    | None ->
      if !mode = `Infer || ty
      then prefix_type_dir name
      else prefix_output_dir name
    | Some n ->
      if !mode = `Compile || !mode = `Infer
      then (output_name := None; n)
      else prefix_output_dir name in
  check_or_create_dir (Filename.dirname name);
  chop_extension_if_any name

let set_mode m =
  if !mode = `Link then
    ( if
      (m = `Shared && !kind <> `ServerOpt) ||
      (m = `Infer && !kind = `Client) ||
      (m = `Interface && !kind = `Client)
      then
        usage ()
      else
        mode := m )
  else
    let args =
      let basic_args = ["-pack"; "-a"; "-c"; "output-obj"] in
      let infer_args = if !kind <> `Client then ["-i"; "-infer"] else [] in
      let shared_args = if !kind <> `ServerOpt then ["-shared"] else [] in
      basic_args @ infer_args @ shared_args
    in
    Printf.eprintf
      "Please specify at most one of %s\n%!"
      (String.concat ", " args);
    exit 1

let get_product_name () = match !output_name with
  | None ->
    Printf.eprintf
      "Please specify the name of the output file, using option -o\n%!";
    exit 1
  | Some name ->
    check_or_create_dir (Filename.dirname name);
    name

let build_library () =
  create_process !compiler (
    ["-a" ; "-o"  ; get_product_name () ]
    @ get_common_include ()
    @ !args )

let build_pack () =
  create_process !compiler (
    [ "-pack" ; "-o"  ; get_product_name () ]
    @ get_common_include ()
    @ !args )

let build_obj () =
  create_process !compiler (
    ["-output-obj" ; "-o"  ; get_product_name () ]
    @ get_common_include ()
    @ !args )

let build_shared () =
  create_process !compiler (
    ["-shared" ; "-o"  ; get_product_name () ]
    @ get_common_include ()
    @ !args )

let get_thread_opt () = match !kind with
  | `Client -> []
  | `Server | `ServerOpt -> ["-thread"]

let obj_ext () = if !kind = `ServerOpt then ".cmx" else ".cmo"

(* Process ml and mli files *)

let compile_ocaml ~impl_intf file =
  let obj =
    let ext = match impl_intf with
      | `Impl -> obj_ext ()
      | `Intf -> ".cmi"
    in
    output_prefix file ^ ext in
  create_process !compiler (
    ["-c" ;
     "-o" ; obj ]
    @ preprocess_opt ~ocaml:true !ppopt
    @ !args
    @ get_thread_opt ()
    @ get_common_include ()
    @ get_common_ppx ()
    @ [impl_intf_opt impl_intf; file] )

let output_ocaml_interface file =
  create_process !compiler (
    [ "-i" ]
    @ preprocess_opt ~ocaml:true !ppopt
    @ !args
    @ get_common_include ()
    @ get_common_ppx ()
    @ [file] )

let process_ocaml ~impl_intf file =
  if do_compile () then
    compile_ocaml ~impl_intf file;
  if !mode = `Interface then
    output_ocaml_interface file

let compile_obj file =
  if do_compile () then
    ( create_process !compiler (!args @ [file]);
      args := !args @ [output_prefix file ^ ext_obj] )

(* Process eliom and eliomi files *)

let run_command s =
  let v = Sys.command s in
  if v != 0 then
    failwith (Printf.sprintf "Warning: command [%s] returned %d" s v)

(* WARNING: if you change this, also change inferred_type_prefix in
   ppx/ppx_eliom_utils.ml and ocamlbuild/ocamlbuild_eliom.ml *)
let inferred_type_prefix = "eliom_inferred_type_"

(* FIXME!

   run_sed is a temporary hack to parse weakly monomorphic types of
   the following forms:

   '_a
   _[< ... ]
   _[> ... ]

   It also removes type indices foo/2 (in case of ambiguity in
   module interfaces).

   These appear in type_mli files, but they are not accepted by
   the OCaml parser.  *)
let run_sed file =
  run_command ("sed -i -e 's$/[1-9][0-9]*$$g' " ^ file);
  run_command ("sed -i -e 's/_\\[\\([<>]\\)/[\\1/g' " ^ file);
  run_command
    (Printf.sprintf
       "sed -i -e \"s/'\\(_[a-z0-9_]*\\)/'%s\\1/g\" %s"
       inferred_type_prefix file);
  (* For some sed implementations, [-e] is interpreted as an argument
     to [-i], resulting in copy of the original file named [file ^
     "-e"]. If we find such a file, remove it. Slightly less messy
     than doing OS detection beforehand. *)
  let file_e = file ^ "-e" in
  if Sys.file_exists file_e then Sys.remove file_e

let compile_server_type_eliom file =
  let obj = output_prefix ~ty:true file ^ !server_types_file_ext
  and ppopts = !ppopt @ (if !pp_mode = `Ppx then [] else ["-impl"]) in
  if !do_dump then begin
    let camlp4, ppopt =
      get_pp_dump ["eliom.syntax.type"] ("-printer" :: "o" :: ppopts @ [file]) in
    create_process camlp4 ppopt;
    exit 0
  end;
  let out = Unix.openfile obj [Unix.O_WRONLY; Unix.O_CREAT; Unix.O_TRUNC] 0o666 in
  let on_error _ =
    Unix.close out;
    Sys.remove obj
  in
  create_process ~out ~on_error !compiler (
    [ "-i" ]
    @ preprocess_opt ~kind:`Types ppopts
    @ !args
    @ get_common_include ()
    @ get_common_ppx ()
    @ ["-impl"; file] );
  Unix.close out;
  if !pp_mode = `Ppx then run_sed obj

let output_eliom_interface ~impl_intf file =
  if !do_dump then begin
    Printf.eprintf "Dump (-dump) not supported for interface inference (-i).";
    exit 1
  end;
  let indent ch =
    let spaces =
      match !pp_mode with
      | `Ppx -> ""
      | `Camlp4 -> "  "
    in
    try
      while true do
        let line = input_line ch in
        Printf.printf "%s%s\n" spaces line
      done
    with End_of_file -> ()
  in
  let args kind =
    let ppopts = get_ppopts ~impl_intf file in
    [ "-i" ]
    @ preprocess_opt ~kind ppopts
    @ [ "-intf-suffix"; ".eliomi" ]
    @ !args
    @ get_common_include ~kind ()
    @ get_common_ppx ~kind ()
    @ [ impl_intf_opt impl_intf; file ]
  and open_block str =
    match !pp_mode with
    | `Ppx ->
      Printf.printf "[%%%%%s.start]\n" str
    | `Camlp4 ->
      Printf.printf "{%s{\n" str
  and close_block () =
    match !pp_mode with
    | `Ppx ->
      ()
    | `Camlp4 ->
      print_endline "}}"
  in
  Printf.printf
    "(* WARNING generated in an ad-hoc fashion. Use with care! *)\n";
  open_block "server";
  create_filter !compiler (args `Server) indent;
  close_block ();
  open_block "client";
  create_filter !compiler (args `Client) indent;
  close_block ()

let compile_eliom ~impl_intf file =
  let obj =
    let ext = match !kind with
      | `Client -> ".cmo"
      | `Server | `ServerOpt -> obj_ext ()
    in
    output_prefix file ^ ext
  in
  let ppopts = get_ppopts ~impl_intf file in
  (* if !do_dump then begin *)
  (*   let camlp4, ppopt = get_pp_dump pkg ("-printer" :: "o" :: ppopts @ [file]) in *)
  (*   create_process camlp4 ppopt; *)
  (*   exit 0 *)
  (* end; *)
  create_process !compiler (
    [ "-c" ; "-o"  ; obj ]
    @ preprocess_opt ppopts
    @ [ "-intf-suffix"; ".eliomi" ]
    @ get_thread_opt ()
    @ !args
    @ get_common_include ()
    @ get_common_ppx ()
    @ [impl_intf_opt impl_intf; file] );
  args := !args @ [obj]

let process_eliom ~impl_intf file =
  match !mode with
  | `Infer when impl_intf = `Impl ->
    compile_server_type_eliom file
  | `Interface ->
    output_eliom_interface ~impl_intf file
  | _ ->
    compile_eliom ~impl_intf file

let build_server ?(name = "a.out") () =
  fail "Linking eliom server is not yet supported"
(* TODO ? Build a staticaly linked ocsigenserver. *)

let build_client () =
  let name = chop_extension_if_any (get_product_name ()) in
  let exe = prefix_output_dir (Filename.basename name) in
  check_or_create_dir (Filename.dirname exe);
  let js = name ^ ".js" in
  create_process !compiler (
    ["-o"  ;  exe ]
    @ link_all ()
    @ get_common_include ()
    @ get_client_lib ()
    @ !args );
  create_process !js_of_ocaml (
    ["-o" ; js ]
    @ get_client_js ()
    @ !jsopt
    @ [exe] )

let process_option () =
  let i = ref 1
  and ppx = ref false in
  while !i < Array.length Sys.argv do
    match Sys.argv.(!i) with
    | "-help" | "--help" -> usage ()
    | "-no-autoload" -> autoload_predef := false; incr i
    | "-type-conv" -> type_conv := true; incr i
    | "-dont-force-linkall"->
      if !kind <> `Client then usage ();
      force_link_all := false; incr i
    | "-i" -> set_mode `Interface; incr i
    | "-c" -> set_mode `Compile; incr i
    | "-a" -> set_mode `Library; incr i
    | "-pack" -> set_mode `Pack; incr i
    | "-output-obj" -> set_mode `Obj; incr i
    | "-shared" -> set_mode `Shared; incr i
    | "-infer" -> set_mode `Infer; incr i
    | "-verbose" -> verbose := true; args := !args @ ["-verbose"] ;incr i
    | "-vmthread" -> Printf.eprintf "The -vmthread option isn't supported yet."; exit 1
    | "-o" ->
      if !i+1 >= Array.length Sys.argv then usage ();
      output_name := Some Sys.argv.(!i+1);
      i := !i+2
    | "-dump" ->
      do_dump := not !do_dump;
      i := !i+1
    | "-ppx" when not !ppx ->
      ppx := true;
      pp_mode := `Ppx;
      i := !i+1
    | "-dir" ->
      if !i+1 >= Array.length Sys.argv then usage ();
      build_dir := Sys.argv.(!i+1);
      i := !i+2
    | "-type-dir" ->
      if !i+1 >= Array.length Sys.argv then usage ();
      type_dir := Sys.argv.(!i+1);
      i := !i+2
    | "-server-types-ext" ->
      if !i+1 >= Array.length Sys.argv then usage ();
      server_types_file_ext := Sys.argv.(!i+1);
      i := !i+2
    | "-package" ->
      if !i+1 >= Array.length Sys.argv then usage ();
      package := !package @ split ',' Sys.argv.(!i+1);
      i := !i+2
    | "-predicates" ->
      if !i+1 >= Array.length Sys.argv then usage ();
      predicates := !predicates @ split ',' Sys.argv.(!i+1);
      i := !i+2
    | "-pp" ->
      if !i+1 >= Array.length Sys.argv then usage ();
      pp := Some Sys.argv.(!i+1);
      i := !i+2
    | "-jsopt" ->
      if !kind <> `Client then usage ();
      if !i+1 >= Array.length Sys.argv then usage ();
      jsopt := !jsopt @ [Sys.argv.(!i+1)];
      i := !i+2
    | "-ppopt" ->
      if !i+1 >= Array.length Sys.argv then usage ();
      ppopt := !ppopt @ [Sys.argv.(!i+1)];
      i := !i+2
    | "-intf" ->
      if !i+1 >= Array.length Sys.argv then usage ();
      process_eliom ~impl_intf:`Intf Sys.argv.(!i+1);
      i := !i+2
    | "-impl" ->
      if !i+1 >= Array.length Sys.argv then usage ();
      process_eliom ~impl_intf:`Impl Sys.argv.(!i+1);
      i := !i+2
    | arg when Filename.check_suffix arg ".mli" ->
      process_ocaml ~impl_intf:`Intf arg;
      incr i
    | arg when Filename.check_suffix arg ".ml" ->
      process_ocaml ~impl_intf:`Impl arg;
      incr i
    | arg when Filename.check_suffix arg ".eliom" ->
      process_eliom ~impl_intf:`Impl arg;
      incr i
    | arg when Filename.check_suffix arg ".eliomi" ->
      process_eliom ~impl_intf:`Intf arg;
      incr i
    | arg when Filename.check_suffix arg ".c" ->
      compile_obj arg; incr i
    | arg -> args := !args @ [arg]; incr i
  done;
  match !mode with
  | `Library -> build_library ()
  | `Pack -> build_pack ()
  | `Obj -> build_obj ()
  | `Shared -> build_shared ()
  | `Link when !kind = `Client -> build_client ()
  | `Link (* Server and ServerOpt *) -> build_server ?name:(!output_name) ()
  | `Compile | `Infer | `Interface -> ()

let main () =
  let cmd = Filename.basename Sys.argv.(0) in
  let cmd =
    try
      let idx = String.index cmd '.' in
      String.sub cmd 0 idx
    with Not_found -> cmd in
  let k =
    match cmd with
    | "eliomopt" ->
      compiler := ocamlopt;
      build_dir := default_server_dir;
      `ServerOpt
    | "eliomcp" ->
      compiler := ocamlcp;
      build_dir := default_server_dir;
      `Server
    | "js_of_eliom" ->
      compiler := ocamlc;
      build_dir := default_client_dir;
      `Client
    | "eliomc" ->
      compiler := ocamlc;
      build_dir := default_server_dir;
      `Server
    | s ->
      Format.eprintf "exec name not recognize %S: fallback to ocamlc@." s;
      compiler := ocamlc;
      build_dir := default_server_dir;
      `Server
  in
  kind := k;
  process_option ()

let _ = main ()
